cpp/src/fil/infer.cu

```diff
@@ -160,6 +198,130 @@ struct tree_aggregator_t {
+                                               void* shared_workspace, size_t)
+    : tmp_storage(shared_workspace), num_classes(num_classes_) {}
+
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, float> single_tree_prediction, int tree) {
+    acc += single_tree_prediction;
+  }
+
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int num_outputs) {
+    __syncthreads();  // free up input row
+    // load margin into shared memory
+    best_margin_label<NITEMS> best;
+    vec<NITEMS, float>* per_thread = (vec<NITEMS, float>*)tmp_storage;
+    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+
+    __syncthreads();
+    // reduce per-thread margin summand into per-class complete margin
+    // (for each of the NITEMS rows)
+    if (threadIdx.x < num_classes) {
+      for (int c = threadIdx.x + num_classes; c < blockDim.x; c += num_classes)
+        acc += per_thread[c];
+      best.margin = acc;
+      best.label.fill(threadIdx.x);
+    }
+    __syncthreads();
+    // determine best class per block (for each of the NITEMS rows)
+    typedef BlockReduceMultiClass<NITEMS> BlockReduceT;
+    best = BlockReduceT(*(typename BlockReduceT::TempStorage*)tmp_storage)
+             .Reduce(best, ArgMax<NITEMS>(), num_classes);
+    // write it out to global memory
+    if (threadIdx.x == 0) {
+      for (int i = 0; i < NITEMS; ++i) {
+        int row = blockIdx.x * NITEMS + i;
+        if (row < num_rows) out[row] = best.label[i];
+      }
+    }
+  }
+};
+
+template <int NITEMS>
+struct tree_aggregator_t<NITEMS, TREE_PER_CLASS_MANY_CLASSES> {
+  vec<NITEMS, float> acc;
+  void* tmp_storage;
+  vec<NITEMS, float>* per_class_margin;
+  int num_classes;
+
+  static size_t smem_finalize_footprint(int num_classes) {
+    size_t phase1 = num_classes * sizeof(vec<NITEMS, float>);
+
+    size_t phase2 =
+      sizeof(typename BlockReduceHostMultiClass<NITEMS>::TempStorage);
+    return std::max(phase1, phase2);
+  }
+
+  static size_t smem_accumulate_footprint(int num_classes) {
+    return num_classes * sizeof(vec<NITEMS, float>);
+  }
+
+  __device__ __forceinline__ tree_aggregator_t(int num_classes_,
+                                               void* shared_workspace,
+                                               size_t data_row_size)
+    : tmp_storage(shared_workspace),
+      per_class_margin(
+        (vec<NITEMS, float>*)((char*)shared_workspace + data_row_size)),
+      num_classes(num_classes_) {
+    for (int c = threadIdx.x; c < num_classes; c += blockDim.x)
+      per_class_margin[c].fill(0.0f);
+    //__syncthreads(); // done in the main loop
+  }
+
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, float> single_tree_prediction, int tree) {
+    // since threads are assigned to consecutive classes, no need for atomics
+    per_class_margin[tree % num_classes] += single_tree_prediction;
+    //__syncthreads(); // done in the main loop
```

Comment:
`// __syncthreads() is performed in the main loop` (to avoid confusion with commented-out code)